test/jasmine/tests/axes_test.js

```diff
@@ -321,15 +321,137 @@ describe('Test axes', function() {
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('trace');
+            });
+
+            it('should not switch categorymode on "array" if categorylist is supplied but empty', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorylist: []}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('trace');
+            });
+        });
+
+        describe('do NOT set categorymode to "array" if it has some other proper value', function() {
+
+            it('should use specified categorymode if it is supplied even if categorylist exists', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'trace', categorylist: ['b','a','d','e','c']}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('trace');
+            });
+
+            it('should use specified categorymode if it is supplied even if categorylist exists', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'category ascending', categorylist: ['b','a','d','e','c']}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('category ascending');
+            });
+
+            it('should use specified categorymode if it is supplied even if categorylist exists', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'category descending', categorylist: ['b','a','d','e','c']}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('category descending');
+            });
+
+        });
+
+        describe('setting categorymode to the default if the value is unexpected', function() {
+
+            it('should switch categorymode to "trace" if mode is supplied but invalid', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'invalid value'}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('trace');
+            });
+
+            it('should switch categorymode to "array" if mode is supplied but invalid and list is supplied', function() {
+                PlotlyInternal.plot(gd, [{x: ['c','a','e','b','d'], y: [15,11,12,13,14]}], {
+                    xaxis: {type: 'category', categorymode: 'invalid value', categorylist: ['b','a','d','e','c']}
+                });
+                expect(gd._fullLayout.xaxis.categorymode).toBe('array');
+            });
+
+        });
+
+    });
+
```

Comment:
@etpinard May I ask you to take a look at these test cases if they seem to be in line with what you wrote about categorymode coercion earlier.